core/services/functions/connector_handler.go

```diff
@@ -25,35 +29,60 @@ import (
 type functionsConnectorHandler struct {
 	services.StateMachine
 
-	connector      connector.GatewayConnector
-	signerKey      *ecdsa.PrivateKey
-	nodeAddress    string
-	storage        s4.Storage
-	allowlist      functions.OnchainAllowlist
-	rateLimiter    *hc.RateLimiter
-	subscriptions  functions.OnchainSubscriptions
-	minimumBalance assets.Link
-	lggr           logger.Logger
+	connector           connector.GatewayConnector
+	signerKey           *ecdsa.PrivateKey
+	nodeAddress         string
+	storage             s4.Storage
+	allowlist           functions.OnchainAllowlist
+	rateLimiter         *hc.RateLimiter
+	subscriptions       functions.OnchainSubscriptions
+	minimumBalance      assets.Link
+	listener            FunctionsListener
+	offchainTransmitter OffchainTransmitter
+	heartbeatRequests   map[RequestID]*HeartbeatResponse
+	orderedRequests     []RequestID
+	mu                  sync.Mutex
+	chStop              chan struct{}
+	shutdownWaitGroup   sync.WaitGroup
+	serviceContext      context.Context
+	serviceCancel       context.CancelFunc
+	lggr                logger.Logger
 }
 
+const (
+	HeartbeatRequestTimeoutSec = 240
+	HeartbeatCacheSize         = 1000
+)
+
 var (
 	_ connector.Signer                  = &functionsConnectorHandler{}
 	_ connector.GatewayConnectorHandler = &functionsConnectorHandler{}
 )
 
-func NewFunctionsConnectorHandler(nodeAddress string, signerKey *ecdsa.PrivateKey, storage s4.Storage, allowlist functions.OnchainAllowlist, rateLimiter *hc.RateLimiter, subscriptions functions.OnchainSubscriptions, minimumBalance assets.Link, lggr logger.Logger) (*functionsConnectorHandler, error) {
-	if signerKey == nil || storage == nil || allowlist == nil || rateLimiter == nil || subscriptions == nil {
-		return nil, fmt.Errorf("signerKey, storage, allowlist, rateLimiter and subscriptions must be non-nil")
+// internal request ID is a hash of (sender, requestID)
+func InternalId(sender []byte, requestId []byte) RequestID {
+	idPair := sender
+	idPair = append(idPair, requestId...)
```

Comment:
`append` does not modify the first argument, see my example here https://goplay.tools/snippet/4b9XBSac8qH

Reply:
I assign back to "idPair" so it is modified, right?

Reply:
What I meant, why not simplify:
```
idPair := append(sender, requestId...)
```